File: cpp/031-040/Search for a Range.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        std::vector<int>::iterator low,up;
        low =std::lower_bound (nums.begin(), nums.end(), target);
        up = std::upper_bound (nums.begin(), nums.end(), target); 
        vector<int> result;
        if (low == up) {
            result.push_back(-1);
            result.push_back(-1);
        } else {
            result.push_back(low - nums.begin());
            result.push_back(up - nums.begin() - 1);
        }
        return result;
    }
};



class Solution {
    
    int FindFirst(const vector<int>& nums, int target) {
        int i = 0;
        int j = nums.size() - 1;
        
        int pos = -1;
        while (i <= j) {
            int mid = (i + j) / 2;
            
            if (nums[mid] == target) {
                pos = mid;
                j = mid - 1;
            } else if (nums[mid] > target) {
                j = mid - 1;
            } else {
                i = mid + 1;
            }
        }
        
        return pos;
    }
    
    int FindLast(const vector<int>& nums, int target) {
        int i = 0;
        int j = nums.size() - 1;
        
        int pos = -1;
        while (i <= j) {
            int mid = (i + j) / 2;
            
            if (nums[mid] == target) {
                pos = mid;
                i = mid + 1;
            } else if (nums[mid] > target) {
                j = mid - 1;
            } else {
                i = mid + 1;
            }
        }
        
        return pos;
    }
    
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> ans{FindFirst(nums, target), FindLast(nums, target)};
        return ans;
    }
};

File: python/31-40/Search for a Range.py
class Solution:
    # @param {integer[]} nums
    # @param {integer} target
    # @return {integer[]}
    def searchRange(self, nums, target):
        low = 0
        high = len(nums)-1
        res = []
        
        while low <= high:
            mid = (low+high)/2
            if(nums[mid] < target):
                low = mid+1
            elif(nums[mid] > target):
                high = mid-1
            else:
                bound1 = bound2 = mid
                while bound1 >= 0 and nums[bound1] == nums[mid]:
                    bound1 -= 1
                while bound2 < len(nums) and nums[bound2] == nums[mid]:
                    bound2 += 1
                res.append(bound1+1)
                res.append(bound2-1)
                break
        
        if(len(res) == 0):
            res.append(-1)
            res.append(-1)
            
        return res
